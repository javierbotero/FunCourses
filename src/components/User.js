import React, { useEffect } from 'react';
import ReactRouterPropTypes from 'react-router-prop-types';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import styles from '../css/User.module.css';
import course from '../css/Course.module.css';
import avatarImage from '../assets/images/avatar.jpg';

const User = props => {
  const {
    courses,
    user,
    match,
    location,
    url,
    coursesToDivs,
    commentsToDivsWithCourse,
    handleCreateFriendship,
    tokenPayload,
    userPayload,
    id,
    token,
    urlApi,
    objThunk,
    isFriendshipRequested,
    handleDelFriend,
    findCourses,
    setShowMenu,
    userId,
    userPassword,
  } = props;
  useEffect(() => {
    const deskTopMenu = document.querySelector('.desktopMenu');
    if (deskTopMenu.classList.contains('hide')) {
      deskTopMenu.classList.remove('hide');
    }
  });
  const foundUser = (users, id) => users.find(u => u.id === id);
  const selectUser = (foundUser, user, match, location) => {
    const found = foundUser([...user.requests, ...user.pendings], parseInt(match.params.id, 10));
    let result;
    if (found) {
      result = found;
    } else if (user.id === parseInt(match.params.id, 10)) {
      result = user;
    } else {
      result = location.state.user;
    }
    return result;
  };
  const infoUser = selectUser(foundUser, user, match, location);
  const infoUserToHtml = (obj, courses, findCoursesCb) => {
    if (obj.status) {
      return (
        <div className={styles.friendTrue}>
          <div
            style={{
              backgroundImage: obj.url_avatar ? `url(${urlApi}${obj.url_avatar.slice(1, obj.url_avatar.length)})` : `url(${avatarImage})`,
            }}
            className={`${course.mainPic}`}
          >
            <div className={course.degrade} />
            <div className={course.infoTeacher}>
              <div>
                {obj.status}
              </div>
            </div>
            <div className={course.price}>
              <h3>{obj.username}</h3>
            </div>
          </div>
          <h3 className={styles.titleCourses}>Courses</h3>
          <div className={styles.interactions}>
            <h4>Student</h4>
            <div className={styles.courses}>
              {coursesToDivs(findCoursesCb(courses, obj.courses_as_student), url, location)}
            </div>
          </div>
          <div className={styles.interactions}>
            <h4>Teacher</h4>
            <div className={styles.courses}>
              {coursesToDivs(findCoursesCb(courses, obj.courses), url, location)}
            </div>
          </div>
          <div className={styles.comments}>
            <h4 className={course.commentsTitle}>
              {infoUser.username}
              &apos;s opinions
            </h4>
            {commentsToDivsWithCourse(obj.comments, location, url)}
          </div>
        </div>
      );
    }
    return (
      <div className={styles.friendFalse}>
        <div
          style={{
            backgroundImage: infoUser.url_avatar ? `url(${urlApi}${infoUser.url_avatar.slice(1, infoUser.url_avatar.length)})` : `url(${avatarImage})`,
          }}
          className={`${styles.avatar} avatar ${styles.avatarNotFriend}`}
        />
        <h3 className={styles.username}>{infoUser.username}</h3>
        <div className={styles.notFriends}>Your are not friends yet</div>
        <button
          className="button-3"
          type="button"
          onClick={() => {
            const isRequested = isFriendshipRequested(user.pending_requested_friendships, obj.id);
            if (isRequested) {
              handleDelFriend(
                tokenPayload,
                userPayload,
                id,
                token,
                userId,
                userPassword,
                urlApi,
                objThunk,
                {
                  id: isRequested.id,
                  pendingRequested: true,
                },
              );
            } else {
              handleCreateFriendship(
                tokenPayload,
                userPayload,
                id,
                token,
                userId,
                userPassword,
                obj.id,
                obj.username,
                objThunk,
                urlApi,
              );
            }
          }}
        >
          { isFriendshipRequested(user.pending_requested_friendships, obj.id) ? 'Cancel request' : 'Add to friend' }
        </button>
      </div>
    );
  };

  return (
    <div>
      <nav className="navMenu">
        <div>
          <Link to={location.state ? location.state.from.pathname : `${url}`}>
            <FontAwesomeIcon icon="chevron-left" />
          </Link>
        </div>
        <div>
          {infoUser.username}
        </div>
        <button className="removeButtonStyles" type="button" onClick={() => setShowMenu(true)}>
          <FontAwesomeIcon icon="ellipsis-v" />
        </button>
      </nav>
      <div className={styles.container}>
        {infoUserToHtml(infoUser, courses, findCourses)}
      </div>
    </div>
  );
};

User.propTypes = {
  match: ReactRouterPropTypes.match.isRequired,
  location: ReactRouterPropTypes.location.isRequired,
  url: PropTypes.string.isRequired,
  user: PropTypes.shape({
    id: PropTypes.number.isRequired,
    username: PropTypes.string.isRequired,
    courses_as_student: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      content: PropTypes.string.isRequired,
      teacher_id: PropTypes.number.isRequired,
      status: PropTypes.string.isRequired,
      dates: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      created_at: PropTypes.string.isRequired,
      updated_at: PropTypes.string.isRequired,
      teacher: PropTypes.shape({
        id: PropTypes.number,
        username: PropTypes.string,
      }).isRequired,
      favorites: PropTypes.arrayOf(PropTypes.shape({
        course_id: PropTypes.number,
        user_id: PropTypes.number,
      })).isRequired,
      subscriptions: PropTypes.arrayOf(PropTypes.shape({
        course_id: PropTypes.number,
        user_id: PropTypes.number.isRequired,
        confirmed: PropTypes.bool,
      })).isRequired,
      comments: PropTypes.arrayOf(PropTypes.shape({
        user_id: PropTypes.number,
        course_id: PropTypes.number,
        body: PropTypes.string,
      })).isRequired,
      pendings: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number,
        user_id: PropTypes.number,
        course_id: PropTypes.number,
        confirmed: PropTypes.bool,
      })).isRequired,
    })).isRequired,
    pending_courses_as_student: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      content: PropTypes.string.isRequired,
      teacher_id: PropTypes.number.isRequired,
      dates: PropTypes.string.isRequired,
    })).isRequired,
    comments: PropTypes.arrayOf(PropTypes.shape({
      body: PropTypes.string.isRequired,
      course_id: PropTypes.number.isRequired,
    })).isRequired,
    favorites: PropTypes.arrayOf(PropTypes.shape({
      course_id: PropTypes.number,
    })).isRequired,
    pending_to_accept_friendships: PropTypes.arrayOf(PropTypes.shape({
      receiver_id: PropTypes.number,
      sender_id: PropTypes.number,
    })).isRequired,
    pending_requested_friendships: PropTypes.arrayOf(PropTypes.shape({
      receiver_id: PropTypes.number,
      sender_id: PropTypes.number,
    })).isRequired,
    friendship_requests: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number,
      username: PropTypes.string,
      email: PropTypes.string,
    })).isRequired,
    requests: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number,
        status: PropTypes.string,
        email: PropTypes.string,
        username: PropTypes.string,
        courses_as_student: PropTypes.arrayOf({
          id: PropTypes.number,
          title: PropTypes.string,
          content: PropTypes.string,
          favorites: PropTypes.arrayOf(PropTypes.shape({
            id: PropTypes.number,
            user_id: PropTypes.number,
            course_id: PropTypes.number,
          })),
        }),
      }),
    ).isRequired,
    pendings: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number,
        status: PropTypes.string,
        email: PropTypes.string,
        username: PropTypes.string,
        courses_as_student: PropTypes.arrayOf({
          id: PropTypes.number,
          title: PropTypes.string,
          content: PropTypes.string,
          favorites: PropTypes.arrayOf(PropTypes.shape({
            id: PropTypes.number,
            user_id: PropTypes.number,
            course_id: PropTypes.number,
          })),
        }),
      }),
    ).isRequired,
    courses: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      content: PropTypes.string.isRequired,
      teacher_id: PropTypes.number.isRequired,
      status: PropTypes.string.isRequired,
      dates: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      created_at: PropTypes.string.isRequired,
      updated_at: PropTypes.string.isRequired,
      teacher: PropTypes.shape({
        id: PropTypes.number,
        username: PropTypes.string,
      }).isRequired,
      favorites: PropTypes.arrayOf(PropTypes.shape({
        course_id: PropTypes.number,
        user_id: PropTypes.number,
      })).isRequired,
      subscriptions: PropTypes.arrayOf(PropTypes.shape({
        course_id: PropTypes.number,
        user_id: PropTypes.number.isRequired,
        confirmed: PropTypes.bool,
      })).isRequired,
      comments: PropTypes.arrayOf(PropTypes.shape({
        user_id: PropTypes.number,
        course_id: PropTypes.number,
        body: PropTypes.string,
      })),
      pendings: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number,
        user_id: PropTypes.number,
        course_id: PropTypes.number,
        confirmed: PropTypes.bool,
      })).isRequired,
    })),
  }).isRequired,
  courses: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    content: PropTypes.string.isRequired,
    teacher_id: PropTypes.number.isRequired,
    status: PropTypes.string.isRequired,
    dates: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    created_at: PropTypes.string.isRequired,
    updated_at: PropTypes.string.isRequired,
    teacher: PropTypes.shape({
      id: PropTypes.number,
      username: PropTypes.string,
    }).isRequired,
    favorites: PropTypes.arrayOf(PropTypes.shape({
      course_id: PropTypes.number,
      user_id: PropTypes.number,
    })).isRequired,
    subscriptions: PropTypes.arrayOf(PropTypes.shape({
      course_id: PropTypes.number,
      user_id: PropTypes.number.isRequired,
      confirmed: PropTypes.bool,
    })).isRequired,
    comments: PropTypes.arrayOf(PropTypes.shape({
      user_id: PropTypes.number,
      course_id: PropTypes.number,
      body: PropTypes.string,
    })),
    pendings: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number,
      user_id: PropTypes.number,
      course_id: PropTypes.number,
      confirmed: PropTypes.bool,
    })).isRequired,
  })).isRequired,
  coursesToDivs: PropTypes.func.isRequired,
  commentsToDivsWithCourse: PropTypes.func.isRequired,
  handleCreateFriendship: PropTypes.func.isRequired,
  tokenPayload: PropTypes.func.isRequired,
  userPayload: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
  token: PropTypes.string.isRequired,
  userId: PropTypes.func.isRequired,
  userPassword: PropTypes.func.isRequired,
  objThunk: PropTypes.func.isRequired,
  urlApi: PropTypes.string.isRequired,
  isFriendshipRequested: PropTypes.func.isRequired,
  handleDelFriend: PropTypes.func.isRequired,
  findCourses: PropTypes.func.isRequired,
  setShowMenu: PropTypes.func.isRequired,
};

export default User;
