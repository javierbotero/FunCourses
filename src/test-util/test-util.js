import fetchMock from 'fetch-mock';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { URL } from '../constants/constants';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const mockMatch = {
  params: { id: '5' },
  path: 'some-path',
  url: 'url',
};
const mockCourses = [
  {
    id: 5,
    title: 'Karate class',
    content: 'Nice Karate class',
    teacher_id: 7,
    status: 'Open',
    dates: '2021-03-25T00:00:00.000Z 2021-03-29T00:00:00.000Z ',
    price: 10,
    images_url: [],
    created_at: '2021-03-07T17:56:22.487Z',
    updated_at: '2021-03-15T04:21:24.489Z',
    teacher: {
      id: 7,
      username: 'Adele',
    },
    pendings: [],
    pending_students: [],
    favorites: [],
    subscriptions: [],
    confirmed_students: [],
  },
];
const mockHistory = {
  action: 'PUSH',
  block: jest.fn(),
  createHref: jest.fn(),
  go: jest.fn(),
  listen: jest.fn(),
  push: jest.fn(),
  location: {
    hash: '',
    pathname: '',
    search: '',
  },
  replace: jest.fn(),
};
const mockUser = {
  id: 7,
  status: 'Student',
  username: 'Javier 4',
  url_avatar: 'api.com/users/1',
  courses_as_student: [],
  pending_courses_as_student: [],
  comments: [],
  favorites: [],
  pending_to_accept_friendships: [],
  pending_requested_friendships: [],
  friendship_requests: [],
  requests: [],
  pendings: [],
  courses: [
    {
      id: 5,
      title: 'Karate class',
      content: 'Nice Karate class',
      teacher_id: 7,
      status: 'Open',
      dates: '2021-03-25T00:00:00.000Z 2021-03-29T00:00:00.000Z ',
      price: 10,
      created_at: '2021-03-07T17:56:22.487Z',
      updated_at: '2021-03-15T04:21:24.489Z',
      teacher: {
        id: 7,
        username: 'Adele',
      },
      pendings: [],
      pending_students: [],
      favorites: [],
      subscriptions: [],
      confirmed_students: [],
    },
  ],
};
const mockTeacherCourses = [];
const mockLocation = {
  from: { pathname: 'url' },
  hash: 'some-hash',
  pathname: 'url',
  search: 'default',
  state: {
    user: {
      id: 1,
      username: 'Paco',
    },
    from: {
      pathname: 'url',
    },
  },
};
const mockUrl = 'url';
const mockUsersListToDiv = jest.fn();
const mockCommentsToDivs = jest.fn();
const userPassword = '12345';
const func = jest.fn();
const mockUrlApi = 'url';
const mockToken = 'token';
const mockId = 1;

export {
  func,
  mockStore,
  fetchMock,
  mockMatch,
  mockCourses,
  mockTeacherCourses,
  mockLocation,
  mockUrl,
  mockUsersListToDiv,
  mockCommentsToDivs,
  mockUrlApi,
  mockToken,
  mockId,
  mockUser,
  URL,
  mockHistory,
  userPassword,
};
