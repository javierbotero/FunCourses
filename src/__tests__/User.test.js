import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import * as util from '../test-util/test-util';
import User from '../components/User';

describe('Tests for User', () => {
  it('prints Paco text', () => {
    render(
      <BrowserRouter>
        <User
          courses={util.mockCourses}
          match={util.mockMatch}
          location={util.mockLocation}
          url={util.mockUrl}
          user={util.mockUser}
          coursesToDivs={util.mockCoursesToDivs}
          commentsToDivsWithCourse={util.mockCommentsToDivsWithCourse}
          handleCreateFriendship={util.mockHandleCreateFriendship}
          tokenPayload={util.mockTokenPayload}
          userPayload={util.mockUserPayload}
          id={util.mockId}
          token={util.mockToken}
          useAuth={util.mockUseAuth}
          objThunk={util.mockObjThunk}
          urlApi={util.mockUrlApi}
          isFriendshipRequested={util.mockIsFriendshipRequested}
          handleDelFriend={util.mockHandleDelFriend}
          findCourses={util.mockFindCourses}
        />
      </BrowserRouter>,
    );
    expect(screen.getByText('Paco')).toBeInTheDocument();
  });
  it('prints Your are not friends yet text', () => {
    render(
      <BrowserRouter>
        <User
          courses={util.mockCourses}
          match={util.mockMatch}
          location={util.mockLocation}
          url={util.mockUrl}
          user={util.mockUser}
          coursesToDivs={util.mockCoursesToDivs}
          commentsToDivsWithCourse={util.mockCommentsToDivsWithCourse}
          handleCreateFriendship={util.mockHandleCreateFriendship}
          tokenPayload={util.mockTokenPayload}
          userPayload={util.mockUserPayload}
          id={util.mockId}
          token={util.mockToken}
          useAuth={util.mockUseAuth}
          objThunk={util.mockObjThunk}
          urlApi={util.mockUrlApi}
          isFriendshipRequested={util.mockIsFriendshipRequested}
          handleDelFriend={util.mockHandleDelFriend}
          findCourses={util.mockFindCourses}
        />
      </BrowserRouter>,
    );
    expect(screen.getByText('Your are not friends yet')).toBeInTheDocument();
  });
  it('Prints correctly', () => {
    const tree = renderer.create(
      <BrowserRouter>
        <User
          courses={util.mockCourses}
          match={util.mockMatch}
          location={util.mockLocation}
          url={util.mockUrl}
          user={util.mockUser}
          coursesToDivs={util.mockCoursesToDivs}
          commentsToDivsWithCourse={util.mockCommentsToDivsWithCourse}
          handleCreateFriendship={util.mockHandleCreateFriendship}
          tokenPayload={util.mockTokenPayload}
          userPayload={util.mockUserPayload}
          id={util.mockId}
          token={util.mockToken}
          useAuth={util.mockUseAuth}
          objThunk={util.mockObjThunk}
          urlApi={util.mockUrlApi}
          isFriendshipRequested={util.mockIsFriendshipRequested}
          handleDelFriend={util.mockHandleDelFriend}
          findCourses={util.mockFindCourses}
        />
      </BrowserRouter>,
    );
    expect(tree).toMatchSnapshot();
  });
});
