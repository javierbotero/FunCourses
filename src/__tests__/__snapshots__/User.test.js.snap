// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for User Prints correctly 1`] = `
<div>
  <nav>
    <ul>
      <li>
        <a
          href="/url"
          onClick={[Function]}
        >
          &lt;
        </a>
      </li>
    </ul>
  </nav>
  <h3>
    Paco
  </h3>
  <div>
    <div>
      Your are not friends yet
    </div>
    <button
      onClick={[Function]}
      type="button"
    >
      Add to friend
    </button>
  </div>
</div>
`;
